proxy/proxy.go

```diff
@@ -462,177 +349,213 @@ func (p *Proxy) lookupRoute(r *http.Request) (rt *routing.Route, params map[stri
+		ctx.ensureDefaultResponse()
+	} else if p.flags.Debug() {
+		debugReq, err := mapRequest(ctx.request, ctx.route, ctx.outgoingHost)
+		if err != nil {
+			dbgResponse(ctx.responseWriter, &debugInfo{
+				route:        &ctx.route.Route,
+				incoming:     ctx.OriginalRequest(),
+				response:     &http.Response{StatusCode: http.StatusInternalServerError},
+				err:          err,
+				filterPanics: ctx.debugFilterPanics,
+			})
+
+			return errProxyCanceled
+		}
+
+		ctx.outgoingDebugRequest = debugReq
+		ctx.setResponse(&http.Response{Header: make(http.Header)}, p.flags.PreserveOriginal())
+	} else {
+		backendStart := time.Now()
+		rsp, err := p.makeBackendRequest(ctx)
+		if err != nil {
+			p.metrics.IncErrorsBackend(ctx.route.Id)
+			return err
 		}
 
-		p.metrics.MeasureBackend(rt.Id, start)
-		p.metrics.MeasureBackendHost(rt.Host, start)
-		c.res = rs
+		ctx.setResponse(rsp, p.flags.PreserveOriginal())
+		p.metrics.MeasureBackend(ctx.route.Id, backendStart)
+		p.metrics.MeasureBackendHost(ctx.route.Host, backendStart)
 	}
 
-	start = time.Now()
-	if !c.served && p.flags.PreserveOriginal() {
-		c.originalResponse = cloneResponseMetadata(c.Response())
+	p.applyFiltersToResponse(processedFilters, ctx)
+	return nil
+}
+
+func (p *Proxy) serveResponse(ctx *context) {
+	if p.flags.Debug() {
+		dbgResponse(ctx.responseWriter, &debugInfo{
+			route:        &ctx.route.Route,
+			incoming:     ctx.OriginalRequest(),
+			outgoing:     ctx.outgoingDebugRequest,
+			response:     ctx.response,
+			filterPanics: ctx.debugFilterPanics,
+		})
+
+		return
 	}
-	p.applyFiltersToResponse(processedFilters, c, onErr)
-	p.metrics.MeasureAllFiltersResponse(rt.Id, start)
 
-	if c.res.Body != nil {
-		defer func(body io.Closer) {
-			err := body.Close()
+	start := time.Now()
+	addBranding(ctx.response.Header)
+	copyHeader(ctx.responseWriter.Header(), ctx.response.Header)
+	ctx.responseWriter.WriteHeader(ctx.response.StatusCode)
+	err := copyStream(ctx.responseWriter.(flusherWriter), ctx.response.Body)
+	if err != nil {
+		p.metrics.IncErrorsStreaming(ctx.route.Id)
+		log.Error("error while copying the response stream", err)
+	} else {
+		p.metrics.MeasureResponse(ctx.response.StatusCode, ctx.request.Method, ctx.route.Id, start)
+	}
+}
+
+// http.Handler implementation
+func (p *Proxy) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	ctx := newContext(w, r, p.flags.PreserveOriginal())
+	ctx.startServe = time.Now()
+	err := p.do(ctx)
+
+	defer func() {
+		if ctx.response != nil && ctx.response.Body != nil {
+			err := ctx.response.Body.Close()
 			if err != nil {
 				log.Error("error during closing the response body", err)
 			}
-		}(c.res.Body)
-	}
+		}
+	}()
 
-	if !c.served {
-		if p.flags.Debug() {
-			dbgResponse(w, &debugInfo{
-				route:        &rt.Route,
-				incoming:     c.OriginalRequest(),
-				outgoing:     debugReq,
-				response:     c.Response(),
-				filterPanics: filterPanics})
+	if err != nil {
+		if err == errProxyCanceled {
```

Comment:
from another comment:

errProxyCanceled: "It indicates to the caller that the current proxy mechanism is canceled due to varying reasons. If there were any errors, they are already handled."

i add this to the errProxyCanceled as docs. Does it make sense?

I am also not sure anymore if this is the best approach. Will have a look if it can be handled differently. The reason for this is that i am planning a feature whose implementation triggered this refactor: #274